### PR TITLE
MRG: Link brains with _LinkViewer and plot_linked_brains

### PR DESCRIPTION
**Pull request by GuillaumeFavelier**
This PR adds the `plot_linked_brains()` function which contains a collection of `_TimeViewer` and share their properties. This is still a work in progress and only the `time_slider` is shared for now (we assume the same time array between the brains). This design provides independant apps and focus on what should be 'linked' with very little change. Please note that it can totally change depending on user needs/experience or performance.

Here is an example:

```py
import os
import os
import mne
from mne.datasets import sample
from mne.viz import plot_linked_brains
data_path = sample.data_path()
sample_dir = os.path.join(data_path, 'MEG', 'sample')
subjects_dir = os.path.join(data_path, 'subjects')
fname_stc = os.path.join(sample_dir, 'sample_audvis-meg')
stc1 = mne.read_source_estimate(fname_stc, subject='sample')
stc2 = mne.read_source_estimate(fname_stc, subject='sample')
initial_time = 0.1
with mne.viz.use_3d_backend('pyvista'):
    brain1 = stc1.plot(subjects_dir=subjects_dir, initial_time=initial_time,
                       clim=dict(kind='value', pos_lims=[3, 6, 9]),
                       time_viewer=True, hemi='split', views=['lat', 'med'])
    brain2 = stc2.plot(subjects_dir=subjects_dir, initial_time=initial_time,
                       clim=dict(kind='value', pos_lims=[4, 6, 7]),
                       time_viewer=True)
    plot_linked_brains([brain1, brain2])
```

output

**ToDo**

- [x] Link `playback_speed` and `play`
- [x] Remember widget representation to avoid unnecessary loops: IntSlider, ShowView and SmartSlider (UpdateColorbarScale and BumpColorbarPoints will be done in another PR)
- [x] Rework naming convention to avoid confusion between class method and internal callback
- [x] Add `plot_linked_brains` to `mne/viz/tests/test_3d.py`
- [x] Add `_LinkViewer` to `mne/viz/_brain/tests/test_brain.py`
- [x] Remove the local patch for slider normalized coordinates in multi-view
- [x] Fix visibility issue of the sliders with `toggle_interface`
- [x] Fix inconsistency between UpdateColorbarScale and BumpColorbarPoints

**Known issue**

- [x] The slave brain slider handles are not udpated.

It's an item of #7162

### CODE REVIEW

**Comment by hoechenberger** [comment]
This is just one amazing unit of work you’re doing here!! ❤️❤️❤️

**Comment by GuillaumeFavelier** [comment]
What can I use in the docstring of `plot_linked_brains()` as return type? It's complaining when I say the truth:

```
    """Plot multiple SourceEstimate objects with PyVista.

    Parameters
    ----------
    brains : list, tuple or np.ndarray
        The collection of brains to plot.

    Returns
    -------
    link_viewer : instance of _brain._LinkViewer
       The interactive interface to plot the linked brains.
    """
```
Or is it okay not to specify a return at all?

**Comment by agramfort** [comment]
only certain types of types are valid. Public functions should only link to
public types. One solution is too keep _plot_linked_brain function private

>

**Comment by larsoner** [comment]
Could you just not return anything?

**Comment by larsoner** [comment]
(If there is a return you must describe it, but if you don't return anything then you don't need a returns section at all)

**Comment by GuillaumeFavelier** [comment]
I'll try returning nothing then if it does not work I'll make it private. Thank you.

**Comment by GuillaumeFavelier** [comment]
Now the `play` toggle and the `playback_speed` are linked too:

output

What else could be linked between the windows?

**Comment by GuillaumeFavelier** [comment]
I can take care of `UpdateColorbarScale` and `BumpColorbarPoints` in another PR so this is ready for reviews @larsoner , @agramfort . Feel free to share if you have any feedback. I basically tested locally by duplicating `sample` which is not the best use case so I encourage you to try it out.

**Comment by agramfort** [comment]
Let's see what mne folks think about the public API of this.

option 1:
brain1 = stc.plot(...)
brain2 = stc.plot(...)
mne.viz.link_brains([brain1, brain2])

option 2:
mne.viz.plot_source_estimates([stc1, stc2], ...)

option 3:
mne.viz.plot_stc([stc1, stc2], ...)  # requires a renaming

--
option1 offers more flexibility as it allows to have eg split hemi for
brain1 and not brain2.

thoughts?

**Comment by larsoner** [comment]
I prefer option 1

**Comment by hoechenberger** [comment]
+1 on option 1 from my side too

**Comment by jasmainak** [comment]
I feel option 1 will result in errors being thrown late and spaghetti code, for e.g., if the two brain objects are not compatible and cannot be linked. Option 2 or option 3 looks better in that sense.

**Comment by hoechenberger** [comment]
> I feel option 1 will result in errors being thrown late and spaghetti code, for e.g., if the two brain objects are not compatible and cannot be linked. Option 2 or option 3 looks better in that sense.

In that case, I'd be in favor of option 3 :)

**Comment by larsoner** [comment]
I still prefer the flexibility of doing separate `.plot` calls plus a good error to say why there is an incompatibility (option 1) despite this potential drawback

**Comment by drammock** [comment]
Option 1 seems best to me

**Comment by kingjr** [comment]
I would suggest following the ipywidget API with HBox and VBox, and let the user link the view, sliders etc him/herself.

- scenario 1: compare two effects (audio versus visual and left versus right) each with different signal-to-noise ratios. You want to have the same view and same time but not same colorbar
- scenario 2: compare two views of the same brain (as in the above gif): link everything but the view
- scenario 3: compare multiple subjects: link everything (view, time, colorbars)

**Comment by GuillaumeFavelier** [comment]
I am not so familiar with ipywidget API, could you share an example to see how it would look? Re-reading your comment, it seems there are two variables of interest (view and colorbar since time is always linked).

But I think all the scenarii you mention are possible with option 3 with the right values for `views`, `hemi` and `clim`.

**Comment by GuillaumeFavelier** [comment]
If linked views are something interesting, I can add that to #7162 once we are set on the API.

**Comment by agramfort** [comment]
Let's start with 1 then making options to link as simple as possible to
start. We can then
have later feature requests to extend.

>

**Comment by GuillaumeFavelier** [comment]
> option 1:
brain1 = stc.plot(...)
brain2 = stc.plot(...)
mne.viz.link_brains([brain1, brain2])

```py
    brain1 = stc1.plot(subjects_dir=subjects_dir, initial_time=initial_time,
                       clim=dict(kind='value', pos_lims=[3, 6, 9]),
                       hemi='split', views=['lat', 'med'])
    brain2 = stc2.plot(subjects_dir=subjects_dir, initial_time=initial_time,
                       clim=dict(kind='value', pos_lims=[4, 6, 7]))
    plot_linked_brains([brain1, brain2])
```

The current state of the code already follows the option 1 then. Don't hesitate to tell me if you want me to change anything.

**Comment by agramfort** [comment]
I would rename plot_linked_brains to linked_brains as it does plot but
links existing plots

**Comment by GuillaumeFavelier** [comment]
`linked_brains` or `link_brains` ?

**Comment by larsoner** [comment]
+1 for `link_brains`

**Comment by larsoner** [comment]
Using the modified example:
```
import os
import mne
from mne.datasets import sample
from mne.viz import link_brains
data_path = sample.data_path()
sample_dir = os.path.join(data_path, 'MEG', 'sample')
subjects_dir = os.path.join(data_path, 'subjects')
fname_stc = os.path.join(sample_dir, 'sample_audvis-meg')
stc1 = mne.read_source_estimate(fname_stc, subject='sample')
stc2 = mne.read_source_estimate(fname_stc, subject='sample')
initial_time = 0.1
with mne.viz.use_3d_backend('pyvista'):
    brain1 = stc1.plot(subjects_dir=subjects_dir, initial_time=initial_time,
                       clim=dict(kind='value', pos_lims=[3, 6, 9]),
                       time_viewer=True, hemi='split', views=['lat', 'med'])
    brain2 = stc2.plot(subjects_dir=subjects_dir, initial_time=initial_time,
                       clim=dict(kind='value', pos_lims=[4, 6, 7]),
                       time_viewer=True)
    link_brains([brain1, brain2])
```
I get the second window looking a bit strange:

Screenshot from 2020-01-22 14-59-53

Screenshot from 2020-01-22 14-59-56

There are also rendering problems like adjusting a slider leaves "phantom" sliders sometimes (seems to happen when the clim bumping/min<=mid<=max behavior is activated):

Screenshot from 2020-01-22 15-02-34

And I've gotten some placement problems in multi-view mode:

Screenshot from 2020-01-22 15-02-40

Not sure what's causing these things, but they seem orthogonal to this PR (occur before linking I think). Thus `link_brains` seems to link the time controls well (it does not link the colorbar behaviors). If this is the only intended effect in this PR, then +1 for merge for me and fixing the strange-window behavior and adding colorbar-linking options (somehow) separately.

**Comment by GuillaumeFavelier** [comment]
Oh I forgot! I ported the patch upstream! So everything is in the coordinate system [0, 0.5]. My bad. I can fix it her if you want.

**Comment by agramfort** [comment]
I see indeed UI issues; Duplicated slider buttons for example; Let's look
together tomorrow @GuillaumeFavelier

>

**Comment by GuillaumeFavelier** [comment]
The issue with the duplicate slider handles is new though...

**Comment by GuillaumeFavelier** [comment]
This is caused by the outdated slider representation: It turns out that it is built at every update of the slider so it does not makes sense to store it. I thought it could help improve performance. I will revert.

**Comment by GuillaumeFavelier** [comment]
Not related at all to the representation:  `_TimeViewer` wrapping was applied twice.

> Let's look together tomorrow @GuillaumeFavelier

Sure, my pleasure

File: mne/viz/_brain/_timeviewer.py [hunk 1]
```diff
@@ -225,90 +255,105 @@ def __init__(self, brain):
...
+            plotter=self.plotter,
+            callback=self.set_playback_speed,
+            name="playback_speed"
+        )
         playback_speed_slider = self.plotter.add_slider_widget(
-            self.set_playback_speed,
+            self.playback_speed_call,
             value=default_playback_speed,
-            rng=[0.01, 1], title="playback speed",
+            rng=[0.01, 1], title="play speed",
```
Comment (larsoner) [inline mne/viz/_brain/_timeviewer.py]: We should just use `speed`, it's short but still descriptive. Let's do it in another PR unless you're planning to push more commits

Comment (GuillaumeFavelier) [reply mne/viz/_brain/_timeviewer.py]: let's do it here. it's a very short change

**Comment by GuillaumeFavelier** [comment]
I implemented @larsoner feedbacks and fix the conflicts with `master`.

This is ready to go on my end @agramfort

File: mne/viz/_3d.py [hunk 2]
```diff
@@ -1536,6 +1536,32 @@ def _plot_mpl_stc(stc, subject=None, surface='inflated', hemi='lh',
...
 
+def link_brains(brains):
+    """Plot multiple SourceEstimate objects with PyVista.
+
+    Parameters
+    ----------
+    brains : list, tuple or np.ndarray
+        The collection of brains to plot.
+    """
+    import collections
```
Comment (agramfort) [inline mne/viz/_3d.py]: I would not nest this import

File: mne/viz/_3d.py [hunk 3]
```diff
@@ -1536,6 +1536,32 @@ def _plot_mpl_stc(stc, subject=None, surface='inflated', hemi='lh',
...
+    from ._brain import _Brain, _TimeViewer, _LinkViewer
```
Comment (agramfort) [inline mne/viz/_3d.py]: do we need to do a check that this function requires pyvista backend?

File: mne/viz/_3d.py [hunk 4]
```diff
@@ -1536,6 +1536,32 @@ def _plot_mpl_stc(stc, subject=None, surface='inflated', hemi='lh',
...
+        brains = [brains]
+    if len(brains) == 0:
+        raise ValueError("The collection of brains is empty.")
+    for brain in brains:
+        if isinstance(brain, _Brain):
+            # check if the _TimeViewer wrapping is not already applied
+            if not hasattr(brain, 'time_viewer') or brain.time_viewer is None:
+                brain = _TimeViewer(brain)
+        else:
+            raise TypeError("Expected type is _Brain but"
```
Comment (agramfort) [inline mne/viz/_3d.py]: I would not refer to a private type.
```suggestion
            raise TypeError("Expected type is Brain but"
```
this is now a bit wrong but I prefer this.

File: mne/viz/_brain/_timeviewer.py [hunk 5]
```diff
@@ -131,19 +136,44 @@ def __call__(self, value, update_widget=False):
...
+                    name = getattr(slider, "name", None)
+                    if name == self.name:
+                        self.slider_rep = slider.GetRepresentation()
+            if self.slider_rep is not None:
+                self.slider_rep.SetValue(idx)
+                self.slider_rep.SetTitleText(self.orientation[idx])
+
+
+class SmartSlider(object):
+    """Class to manage smart slider."""
```
Comment (agramfort) [inline mne/viz/_brain/_timeviewer.py]: can you be more explicit. smart in what sense? What does it allow you to do?

**Comment by GuillaumeFavelier** [comment]
> also I still get the bug with the colormap change when pressing y. It switches to all red.
you want to do this in a follow up PR?

+1 to do this in another PR

**Comment by larsoner** [comment]
> +1 to do this in another PR

Yes I still need to do the refactoring of `_limits_to_control_points`. Plenty of other stuff to work on in the meantime and I see this as a really minor use case issue -- it only happens if you are using auto-reset mode *and* you have one-sided data that you've chosen to display with a two-sided colormap, or vice-versa, so probably quite rare in real use cases.

**Comment by larsoner** [comment]
CircleCI is unrelated, thanks @GuillaumeFavelier !
